src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -504,6 +504,13 @@ private sealed class FileSystemEnumerable<T> : IEnumerable<T>
                         }
                         searchPattern = searchPattern.Substring(lastSlash + 1);
                     }
+
+                    // We need to escape any escape characters in the search pattern
+                    searchPattern = searchPattern.Replace(@"\", @"\\");
```

Comment:
with all this string modification here and in NormalizeSearchPattern, maybe it should be a StringBuilder. Not something to do in this PR though.